flake: TestDeploymentCoordinator

### Describe the bug

```
2022-12-15T15:11:39.0852488Z --- FAIL: TestDeploymentCoordinator (70.09s)
2022-12-15T15:11:39.0855832Z     kcp.go:127: shared kcp server will target configuration "/home/runner/work/kcp/kcp/.kcp/admin.kubeconfig"
2022-12-15T15:11:39.0859071Z     workspaces.go:196: Created organization workspace root:e2e-org-znfgz
2022-12-15T15:11:39.0863338Z     workspaces.go:126: Created root:universal workspace root:e2e-org-znfgz:synctargets
2022-12-15T15:11:39.0866448Z     workspaces.go:126: Created root:universal workspace root:e2e-org-znfgz:workload-1
2022-12-15T15:11:39.0871331Z     workspaces.go:126: Created root:universal workspace root:e2e-org-znfgz:workload-2
2022-12-15T15:11:39.0873940Z     util.go:112: Saving test artifacts for test "TestDeploymentCoordinator" under "/tmp/e2e/TestDeploymentCoordinator/3181476435".
2022-12-15T15:11:39.0879096Z     syncer.go:141: Configuring workspace root:e2e-org-znfgz:synctargets for syncing
2022-12-15T15:11:39.0886386Z     kubectl.go:59: running: KUBECONFIG=/tmp/e2e/TestDeploymentCoordinator/3181476435/artifacts/root_e2e-org-znfgz_synctargets.kubeconfig kubectl kcp workload sync east --syncer-image=kind.local/syncer-c2e3073d5026a8f7f2c47a50c16bdbec:a4d77b220ce934e4e12fb443b04f413e518272ebd263465dca7285428a404ec4 --output-file=- --qps=-1 --feature-gates=CustomResourceValidationExpressions=true --api-import-poll-interval=5s --downstream-namespace-clean-delay=2s
2022-12-15T15:11:39.0887729Z     kubectl.go:89: running: kubectl --kubeconfig /home/runner/work/kcp/kcp/.kcp/kind.kubeconfig apply -f -
2022-12-15T15:11:39.0888663Z     syncer.go:589: checking secret kcp-syncer-east-2fu92b3q/kcp-syncer-east-2fu92b3q for annotation kubernetes.io/service-account.name=kcp-syncer-east-2fu92b3q
2022-12-15T15:11:39.0889766Z     syncer.go:589: checking secret kcp-syncer-east-2fu92b3q/kcp-syncer-east-2fu92b3q-token for annotation kubernetes.io/service-account.name=kcp-syncer-east-2fu92b3q
2022-12-15T15:11:39.0890304Z     syncer.go:523: Cluster "east" is Ready
2022-12-15T15:11:39.0890860Z     syncer.go:141: Configuring workspace root:e2e-org-znfgz:synctargets for syncing
2022-12-15T15:11:39.0892748Z     kubectl.go:59: running: KUBECONFIG=/tmp/e2e/TestDeploymentCoordinator/3181476435/artifacts/root_e2e-org-znfgz_synctargets.kubeconfig kubectl kcp workload sync west --syncer-image=kind.local/syncer-c2e3073d5026a8f7f2c47a50c16bdbec:a4d77b220ce934e4e12fb443b04f413e518272ebd263465dca7285428a404ec4 --output-file=- --qps=-1 --feature-gates=CustomResourceValidationExpressions=true --api-import-poll-interval=5s --downstream-namespace-clean-delay=2s
2022-12-15T15:11:39.0893943Z     kubectl.go:89: running: kubectl --kubeconfig /home/runner/work/kcp/kcp/.kcp/kind.kubeconfig apply -f -
2022-12-15T15:11:39.0895020Z     syncer.go:589: checking secret kcp-syncer-west-2vca45dr/kcp-syncer-west-2vca45dr for annotation kubernetes.io/service-account.name=kcp-syncer-west-2vca45dr
2022-12-15T15:11:39.0896034Z     syncer.go:589: checking secret kcp-syncer-west-2vca45dr/kcp-syncer-west-2vca45dr-token for annotation kubernetes.io/service-account.name=kcp-syncer-west-2vca45dr
2022-12-15T15:11:39.0896775Z     syncer.go:523: Cluster "west" is Ready
2022-12-15T15:11:39.0898211Z     syncer.go:523: Cluster "east" is Ready
2022-12-15T15:11:39.0898651Z     syncer.go:523: Cluster "west" is Ready
2022-12-15T15:11:39.0899109Z     deployment_coordinator_test.go:114: Create 2 locations, one for each SyncTargets
2022-12-15T15:11:39.0899864Z     deployment_coordinator_test.go:118: Bind workload workspace 1 to location workspace for the east location
2022-12-15T15:11:39.0900654Z     bind.go:59: Bind workload workspace root:e2e-org-znfgz:synctargets
2022-12-15T15:11:39.0901894Z     kubectl.go:59: running: KUBECONFIG=/tmp/e2e/TestDeploymentCoordinator/3181476435/artifacts/root_e2e-org-znfgz_workload-1.kubeconfig kubectl kcp bind compute root:e2e-org-znfgz:synctargets --location-selectors=region=east
2022-12-15T15:11:39.0979765Z     deployment_coordinator_test.go:126: Bind workload workspace 2 to location workspace for the east location
2022-12-15T15:11:39.0985417Z     bind.go:59: Bind workload workspace root:e2e-org-znfgz:synctargets
2022-12-15T15:11:39.0987934Z     kubectl.go:59: running: KUBECONFIG=/tmp/e2e/TestDeploymentCoordinator/3181476435/artifacts/root_e2e-org-znfgz_workload-2.kubeconfig kubectl kcp bind compute root:e2e-org-znfgz:synctargets --location-selectors=region=east
2022-12-15T15:11:39.1009918Z     deployment_coordinator_test.go:134: Bind workload workspace 1 to location workspace for the west location
2022-12-15T15:11:39.1012365Z     bind.go:59: Bind workload workspace root:e2e-org-znfgz:synctargets
2022-12-15T15:11:39.1016632Z     kubectl.go:59: running: KUBECONFIG=/tmp/e2e/TestDeploymentCoordinator/3181476435/artifacts/root_e2e-org-znfgz_workload-1.kubeconfig kubectl kcp bind compute root:e2e-org-znfgz:synctargets --location-selectors=region=west
2022-12-15T15:11:39.1064738Z     deployment_coordinator_test.go:142: Bind workload workspace 2 to location workspace for the west location
2022-12-15T15:11:39.1065587Z     bind.go:59: Bind workload workspace root:e2e-org-znfgz:synctargets
2022-12-15T15:11:39.1066882Z     kubectl.go:59: running: KUBECONFIG=/tmp/e2e/TestDeploymentCoordinator/3181476435/artifacts/root_e2e-org-znfgz_workload-2.kubeconfig kubectl kcp bind compute root:e2e-org-znfgz:synctargets --location-selectors=region=west
2022-12-15T15:11:39.1067712Z     deployment_coordinator_test.go:150: Get the root compute APIExport Virtual Workspace URL
2022-12-15T15:11:39.1068262Z     deployment_coordinator_test.go:159: Start the Deployment controller
2022-12-15T15:11:39.1085078Z     accessory.go:76: running: /home/runner/work/kcp/kcp/bin/deployment-coordinator --kubeconfig=/home/runner/work/kcp/kcp/.kcp/admin.kubeconfig --context=base --server=https://10.1.0.83:6444/services/apiexport/root:compute/kubernetes
2022-12-15T15:11:39.1086235Z     deployment_coordinator_test.go:213: Create workload in workload workspace "root:e2e-org-znfgz:workload-1", with replicas set to 4
2022-12-15T15:11:39.1087208Z     deployment_coordinator_test.go:226: Wait for the workload in workspace "root:e2e-org-znfgz:workload-1" to be started and available with 4 replicas
2022-12-15T15:11:39.1088337Z     assertions.go:1691: Waiting for condition, but got: Deployment root:e2e-org-znfgz:workload-1/test should have been prepared for transformation by the coordinator for the east syncTarget
2022-12-15T15:11:39.1088919Z     deployment_coordinator_test.go:230: 
2022-12-15T15:11:39.1089328Z         	Error Trace:	util.go:297
2022-12-15T15:11:39.1090045Z         	            				deployment_coordinator_test.go:230
2022-12-15T15:11:39.1090974Z         	            				deployment_coordinator_test.go:261
2022-12-15T15:11:39.1091485Z         	Error:      	Condition never satisfied
2022-12-15T15:11:39.1091905Z         	Test:       	TestDeploymentCoordinator
2022-12-15T15:11:39.1092577Z         	Messages:   	deployment root:e2e-org-znfgz:workload-1/test was not synced
2022-12-15T15:11:39.1093150Z     accessory.go:68: cleanup: ending `deployment-coordinator`
```

### Steps To Reproduce

Mostly reproducible in some github actions

### Expected Behaviour

Test should not flake

### Additional Context

_No response_

--- a/pkg/virtual/apiexport/controllers/apireconciler/apiexport_apireconciler_controller.go
+++ b/pkg/virtual/apiexport/controllers/apireconciler/apiexport_apireconciler_controller.go
@@ -69,6 +69,9 @@ func NewAPIReconciler(
 
 		apiExportLister:  apiExportInformer.Lister(),
 		apiExportIndexer: apiExportInformer.Informer().GetIndexer(),
+		listAPIExports: func(clusterName logicalcluster.Name) ([]*apisv1alpha1.APIExport, error) {
+			return apiExportInformer.Lister().Cluster(clusterName).List(labels.Everything())
+		},
 
 		queue: queue,
 
@@ -121,6 +124,7 @@ type APIReconciler struct {
 
 	apiExportLister  apisv1alpha1listers.APIExportClusterLister
 	apiExportIndexer cache.Indexer
+	listAPIExports   func(clusterName logicalcluster.Name) ([]*apisv1alpha1.APIExport, error)
 
 	queue workqueue.RateLimitingInterface
 
@@ -143,7 +147,7 @@ func (c *APIReconciler) enqueueAPIResourceSchema(obj interface{}, logger logr.Lo
 		runtime.HandleError(err)
 		return
 	}
-	exports, err := c.apiExportLister.Cluster(clusterName).List(labels.Everything())
+	exports, err := c.listAPIExports(clusterName)
 	if err != nil {
 		runtime.HandleError(err)
 		return
@@ -154,13 +158,13 @@ func (c *APIReconciler) enqueueAPIResourceSchema(obj interface{}, logger logr.Lo
 	}
 
 	if len(exports) == 0 {
-		logger.V(2).Info("No kubernetes APIExport found for APIResourceSchema")
+		logger.V(2).Info("No APIExports found")
 		return
 	}
 
 	for _, export := range exports {
 		logger.WithValues("apiexport", export.Name).V(2).Info("Queueing APIExport for APIResourceSchema")
-		c.enqueueAPIExport(obj, logger.WithValues("reason", "APIResourceSchema change", "apiResourceSchema", name))
+		c.enqueueAPIExport(export, logger.WithValues("reason", "APIResourceSchema change", "apiResourceSchema", name))
 	}
 }
 

--- a/pkg/virtual/apiexport/controllers/apireconciler/apiexport_apireconciler_controller_test.go
+++ b/pkg/virtual/apiexport/controllers/apireconciler/apiexport_apireconciler_controller_test.go
@@ -1,0 +1,75 @@
+/*
+Copyright 2023 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apireconciler
+
+import (
+	"testing"
+
+	"github.com/kcp-dev/logicalcluster/v3"
+	"github.com/stretchr/testify/require"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2/ktesting"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+)
+
+func TestEnqueueAPIResourceSchema(t *testing.T) {
+	c := &APIReconciler{
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), ControllerName),
+		listAPIExports: func(clusterName logicalcluster.Name) ([]*apisv1alpha1.APIExport, error) {
+			return []*apisv1alpha1.APIExport{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "export1",
+					},
+				},
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "export2",
+					},
+				},
+			}, nil
+		},
+	}
+
+	schema := &apisv1alpha1.APIResourceSchema{
+		ObjectMeta: metav1.ObjectMeta{
+			Annotations: map[string]string{
+				logicalcluster.AnnotationKey: "mycluster",
+			},
+			Name: "schema1",
+		},
+	}
+
+	logger, _ := ktesting.NewTestContext(t)
+	c.enqueueAPIResourceSchema(schema, logger)
+
+	require.Equal(t, c.queue.Len(), 2)
+
+	// get the queue keys
+	actual := sets.NewString()
+	item, _ := c.queue.Get()
+	actual.Insert(item.(string))
+	item, _ = c.queue.Get()
+	actual.Insert(item.(string))
+
+	expected := sets.NewString("export1", "export2")
+	require.True(t, expected.Equal(actual))
+}

--- a/pkg/virtual/apiexport/controllers/apireconciler/apiexport_apireconciler_reconcile.go
+++ b/pkg/virtual/apiexport/controllers/apireconciler/apiexport_apireconciler_reconcile.go
@@ -276,15 +276,16 @@ func (c *APIReconciler) getSchemasFromAPIExport(ctx context.Context, apiExport *
 	logger := klog.FromContext(ctx)
 	apiResourceSchemas := map[schema.GroupResource]*apisv1alpha1.APIResourceSchema{}
 	for _, schemaName := range apiExport.Spec.LatestResourceSchemas {
-		apiResourceSchema, err := c.apiResourceSchemaLister.Cluster(logicalcluster.From(apiExport)).Get(schemaName)
+		apiExportClusterName := logicalcluster.From(apiExport)
+		apiResourceSchema, err := c.apiResourceSchemaLister.Cluster(apiExportClusterName).Get(schemaName)
 		if err != nil && !apierrors.IsNotFound(err) {
 			return nil, err
 		}
 		if apierrors.IsNotFound(err) {
 			logger.WithValues(
 				"schema", schemaName,
-				"exportNamespace", apiExport.Namespace,
-				"export", apiExport.Name,
+				"exportClusterName", apiExportClusterName,
+				"exportName", apiExport.Name,
 			).V(3).Info("APIResourceSchema for APIExport not found")
 			continue
 		}

